addr_manager.go

```diff
@@ -161,60 +220,218 @@ func (mgr *AddrManager) SetAddrs(p peer.ID, addrs []ma.Multiaddr, ttl time.Durat
 			continue
 		}
 		// re-set all of them for new ttl.
-		addrs := string(addr.Bytes())
+		addrs := addr.String()
 
 		if ttl > 0 {
 			amap[addrs] = expiringAddr{Addr: addr, TTL: exp}
-
+			if err = mgr.storeAddr(p, addr, exp); err != nil {
+				return err
+			}
 			for _, sub := range subs {
 				sub.pubAddr(addr)
 			}
 		} else {
 			delete(amap, addrs)
+			if err = mgr.deleteAddr(p, addr); err != nil {
+				return err
+			}
 		}
 	}
+	return nil
 }
 
 // Addresses returns all known (and valid) addresses for a given
-func (mgr *AddrManager) Addrs(p peer.ID) []ma.Multiaddr {
+func (mgr *AddrManager) Addrs(p peer.ID) ([]ma.Multiaddr, error) {
 	mgr.addrmu.Lock()
 	defer mgr.addrmu.Unlock()
+	return mgr.addrs(p)
+}
 
-	// not initialized? nothing to give.
-	if mgr.addrs == nil {
-		return nil
-	}
-
-	maddrs, found := mgr.addrs[p]
-	if !found {
-		return nil
+func (mgr *AddrManager) addrs(p peer.ID) ([]ma.Multiaddr, error) {
+	maddrs, err := mgr.warmupCache(p)
+	if err != nil {
+		return nil, err
 	}
 
 	now := time.Now()
 	good := make([]ma.Multiaddr, 0, len(maddrs))
-	var expired []string
-	for s, m := range maddrs {
+	var expired []ma.Multiaddr
+	for _, m := range maddrs {
 		if m.ExpiredBy(now) {
-			expired = append(expired, s)
+			expired = append(expired, m.Addr)
 		} else {
 			good = append(good, m.Addr)
 		}
 	}
 
-	// clean up the expired ones.
-	for _, s := range expired {
-		delete(maddrs, s)
+	// clean up the expired ones
+	for _, a := range expired {
+		delete(maddrs, a.String())
+		if err = mgr.deleteAddr(p, a); err != nil {
+			return nil, err
+		}
 	}
-	return good
+	return good, nil
 }
 
 // ClearAddresses removes all previously stored addresses
 func (mgr *AddrManager) ClearAddrs(p peer.ID) {
 	mgr.addrmu.Lock()
 	defer mgr.addrmu.Unlock()
-	mgr.init()
 
-	mgr.addrs[p] = make(addrSet) // clear what was there before
+	mgr.cache.Add(p, make(addrSet))
+	mgr.deleteAddrs(p)
+}
+
+func (mgr *AddrManager) warmupCache(p peer.ID) (addrSet, error) {
+	cached, found := mgr.cache.Get(p)
+	if found {
+		return cached.(addrSet), nil
+	}
+	stored, err := mgr.loadAddrs(p)
+	if err != nil {
+		return nil, err
+	}
+	mgr.cache.Add(p, stored)
+	return stored, nil
+}
+
+func mkExpAddrKey(p peer.ID, addr ma.Multiaddr) ds.Key {
+	return ds.NewKey(peerNamespace).Instance(peer.IDB58Encode(p)).
```

Comment:
Fixed. I wonder though why would base58 be significantly more expensive than any other base encoding?